src/pages/billing/CreateAzureBillingProjectModal.js

```diff
@@ -0,0 +1,151 @@
+import { Ajax } from 'src/libs/ajax'
+import { reportErrorAndRethrow } from 'src/libs/error'
+import { formHint, FormLabel } from 'src/libs/forms'
+import { useCancellation } from 'src/libs/react-utils'
+import * as Utils from 'src/libs/utils'
+import { validate as validateUuid } from 'uuid'
+import validate from 'validate.js'
+
+
+validate.validators.type.types.uuid = value => validateUuid(value)
+validate.validators.type.messages.uuid = 'must be a UUID'
+
+const CreateAzureBillingProjectModal = ({ onSuccess, onDismiss, billingProjectNameValidator }) => {
+  const [billingProjectName, setBillingProjectName] = useState('')
+  const [subscriptionId, setSubscriptionId] = useState('')
+  const [managedApps, setManagedApps] = useState([])
+  const [errorFetchingManagedApps, setErrorFetchingManagedApps] = useState(undefined)
+  const [chosenManagedApp, setChosenManagedApp] = useState('')
+  const [billingProjectNameTouched, setBillingProjectNameTouched] = useState(false)
+  const [subscriptionIdTouched, setSubscriptionIdTouched] = useState(false)
+  const [existing, setExisting] = useState([])
+  const [isBusy, setIsBusy] = useState(false)
+
+  const signal = useCancellation()
+
+  const submit = _.flow(
+    reportErrorAndRethrow('Error creating billing project'),
+    Utils.withBusyState(setIsBusy)
+  )(async () => {
+    try {
+      await Ajax().Billing.createAzureProject(billingProjectName, chosenManagedApp.tenantId, chosenManagedApp.subscriptionId,
+        chosenManagedApp.managedResourceGroupId)
+      onSuccess()
+    } catch (error) {
+      if (error.status === 409) {
+        setExisting(_.concat(billingProjectName, existing))
+      } else {
+        throw error
+      }
+    }
+  })
+
+  // Field validation
+  const billingProjectNameErrors = validate({ billingProjectName }, { billingProjectName: billingProjectNameValidator(existing) })
+  const subscriptionIdErrors = validate({ subscriptionId }, { subscriptionId: { type: 'uuid' } })
+  const isValidSubscriptionId = !!subscriptionId && !subscriptionIdErrors
+
+  useEffect(() => {
+    if (isValidSubscriptionId) {
+      const fetchManagedApps = Utils.withBusyState(setIsBusy,
+        async () => {
+          try {
+            const managedApps = await Ajax(signal).Billing.listAzureManagedApplications(subscriptionId)
+            setManagedApps(managedApps.managedApps)
+            setErrorFetchingManagedApps(false)
+          } catch (obj) {
+            setManagedApps([])
+            setChosenManagedApp('')
+            setErrorFetchingManagedApps(true)
+            // We can't rely on the formatting of the error, so show a generic message but include the error in the console for debugging purposes.
+            const error = await (obj instanceof Response ? obj.text() : obj)
+            console.error(error)
+          }
+        }
+      )
+      fetchManagedApps()
+    }
+  }, [isValidSubscriptionId, subscriptionId, signal])
+
+  const subscriptionIdError = Utils.cond(
```

Comment:
👍 I like the pattern of _not_ exposing backend error messages to the presentation layer.